Get widget value

AnOctopus opened the pull request:
**Description:** get_widget_state prototype

AnOctopus committed:
Add prototype for beta_widget_value()

This allows a user to access the values of widgets by their keys, within callbacks.

AnOctopus committed:
Fix unit tests for new widget key/state behavior

kmcgrady commented on lib/streamlit/widgets.py:
```diff
@@ -73,8 +74,21 @@ def get_widget_value(self, widget_id: str) -> Optional[Any]:
         return getattr(wstate, value_type)
 
+    def get_previous_widget_value(self, widget_id: str) -> Optional[Any]:
+        """Return the previous value of a widget, or None if no value was set."""
```
Probably could be DRY with `_get_widget_value`, but it's not a big deal for the prototype.

AnOctopus replied:
I somehow missed that `_get_widget_value` is 90% of what the methods do. Will be trivial to clean up.

kmcgrady commented on lib/streamlit/widgets.py:
```diff
@@ -158,3 +172,21 @@ def _get_widget_value(state: Optional[WidgetState]) -> Optional[Any]:
+
+    this_session = Server.get_current().get_session_by_id(ctx.session_id)
+    this_widget_state = this_session.get_widget_states()
+    return this_widget_state.get_widget_value(utils._get_widget_id("", None, key))
```
Looks like we need to deserialize too. Tried this in the PR preview.

```
import streamlit as st

state = st.beta_session_state(s=None)

def checkbox(value):
    state.s = st.beta_widget_value("foo")

vals = ["One", "Two", "Three", "Four"]

value = st.select_slider("Number", vals, key="foo")

st.checkbox("Check me", on_change=checkbox)

st.write(value)
st.write(state.s)
```

Shouldn't be too hard but to get the deserializer in the widget state and run it if possible, though I might move that into the Widget State Manager

AnOctopus replied:
Had trouble parsing your comment but after running the code locally I understand, and have a rough idea how to fix it. 

Definitely not in scope for this prototype but having a clear marker of serialized vs deserialized values would be very good; possibly as some wrapper class.

kmcgrady replied:
Ah yes. It seems like, it should be in scope for the prototype, right? Essentially, we want `get_widget_value` to return the same value as the return type from the widget itself otherwise, it would seem confusing.

I believe this method in the class will do the trick and just calling it here instead of `get_widget_value`

```python
def get_deserialized_value(self, widget_id: str) -> Optional[Any]:
        raw_widget_value = self.get_widget_value(widget_id)
        deserializer = self._widget_deserializers.get(widget_id, None)
        if deserializer is None:
            return raw_widget_value
        
        return deserializer(raw_widget_value)
```

kmcgrady commented on lib/streamlit/script_runner.py:
```diff
@@ -295,9 +295,14 @@ def _run_script(self, rerun_data):
+                # access the new states using beta_widget_state().
                 self._widgets.set_state(rerun_data.widget_states)
+
+                self._widgets.call_callbacks(rerun_data.widget_states)
```
Curious...I wonder if these are related? Basically `set_state` should trigger `call_callbacks` That would require some investigation, so this is likely to be fine for now.

AnOctopus replied:
Hmm, potentially. The api with my current changes is error-prone due to having some implicit ordering expectations. `set_state` triggering `call_callbacks` might be awkward if you need to set the state without triggering for some reason (and tying such things together feels ugly), but it would be safer/easier to use in the normal case, and feels kinda natural of a thing to do.

kmcgrady commented on lib/tests/streamlit/widgets_test.py:
```diff
@@ -111,6 +111,10 @@ def test_call_callbacks(self):
+        # new state to have been set, so the old state will be stored and used
+        # for the callback checks.
+        widgets.set_state(states)
         widgets.set_state(states)
```
The second `set_state` can be done on line 124 (before the `call_callbacks`) right? That's essentially why it's twice?

AnOctopus replied:
Yes, it just needs to happen before `call_callbacks`. I'll move it to make it clearer.

AnOctopus committed:
Refactor widget state getters to use value parsing function, rearrange test

AnOctopus committed:
Deserialize values in beta_widget_value

kmcgrady commented on lib/streamlit/widgets.py:
```diff
@@ -158,3 +158,25 @@ def _get_widget_value(state: Optional[WidgetState]) -> Optional[Any]:
+    widget_states: WidgetStateManager = this_session.get_widget_states()
+
+    widget_id = utils._get_widget_id("", None, key)
+    deserializer = widget_states._widget_deserializers.get(widget_id, lambda x: x)
```
I forgot something here. I realized one thing that some deserializers are not registerred, this is because we only register deserializers [if there is a callback](https://github.com/streamlit/streamlit/blob/session-state/lib/streamlit/elements/utils.py#L145)

I _think_ all that needs to be done is to add the deserializer in any case, which is hopefully a 1-3 line change.

AnOctopus committed:
Register widget deserializer even when no callback

Should eventually be more fully separated out from the callback code, but this is
fine for a prototype.

kmcgrady commented:
LGTM. Sent a message to @asaini for PR Preview
